csrc/scheduler/mma_utils.cpp

```diff
@@ -49,20 +51,31 @@ bool generateSharedMemoryEpilogueHeuristics(
   const size_t smem_c = (size_t)(gemm_tile.cta_tile.m * gemm_tile.cta_tile.n) *
       dataTypeSize(data_types[2]);
 
+  // NOTE: we can simply add these sizes since they should be integer multiples
+  // of 16 bytes, so they will automatically be aligned.
+  const size_t total_without_smem_epilogue = smem_a + smem_b;
+  // Even if we actually do wind up re-claiming smem_a and smem_b, if we
+  // cannot prove it at this point then we have to assume it will not be
+  // reclaimed.
+  const size_t total_with_smem_epilogue = std::max(
+      smem_a + smem_b,
+      (smem_a_reuse_guaranteed ? 0 : smem_a) +
+          (smem_b_reuse_guaranteed ? 0 : smem_b) + smem_c);
+
   // shortcut where occupancy change is ignored.
   if (ignore_occupancy_drop) {
-    return shared_memory_available >= smem_a + smem_b + smem_c;
+    return shared_memory_available >= total_with_smem_epilogue;
   }
 
   // use additional shared memory for epilogue if occupancy is not changed.
   // occupancy is estimated using register and shared memory usage.
   const auto threads_per_sm = getThreadsPerSMGivenRegPerThread(255);
   const auto blocks_per_sm_by_register = threads_per_sm / threads_per_block;
   const auto blocks_per_sm_without_smem_epilogue = std::min(
-      shared_memory_available / (smem_a + smem_b),
+      shared_memory_available / total_without_smem_epilogue,
       (size_t)blocks_per_sm_by_register);
   const auto blocks_per_sm_with_smem_epilogue = std::min(
-      shared_memory_available / (smem_a + smem_b + smem_c),
+      shared_memory_available / total_with_smem_epilogue,
       (size_t)blocks_per_sm_by_register);
```

Comment:
Should we add a case that, if reuse and no-reuse provides the same occupancy, then we don't promote reuse because this will save a sync?

Reply:
That's a good idea. In that case I'll also need to guard `promoteReuse` with `params.use_smem_epilogue`.

Reply:
Actually, now I understand your comment. Yes we'll need a separate parameter indicating whether to reuse memory or not, in addition to `use_smem_epilogue`. I'll push something in a moment.